File: workers/node/src/bullmq.worker.ts
import { Worker, Queue } from 'bullmq';
import { createBullBoard } from '@bull-board/api';
import { BullMQAdapter } from '@bull-board/api/bullMQAdapter';
import { ExpressAdapter } from '@bull-board/express';
import express from 'express';
import { Client } from 'pg';
import type { ProcessedEvent } from '@mq/shared-node';

const DB_URL = process.env['DATABASE_URL'] ?? 'postgresql://postgres:password@localhost:5432/mq_db';
const REDIS_URL = process.env['REDIS_URL'] ?? 'redis://localhost:6379';
const GROUP_ID = 'inventory-group';
const QUEUE_NAME = 'orders';
const SIMULATE_FAILURE = process.env['SIMULATE_FAILURE'] === 'true';
const BULL_BOARD_PORT = parseInt(process.env['BULL_BOARD_PORT'] ?? '3001', 10);

const redisConnection = { url: REDIS_URL };

async function insertProcessedEvent(
    client: Client,
    event: Omit<ProcessedEvent, 'id' | 'processedAt'>
): Promise<void> {
    const query = `
        INSERT INTO processed_events (event_id, group_id, mq_type, data, latency_ms)
        VALUES ($1, $2, $3, $4, $5)
    `;
    await client.query(query, [
        event.eventId,
        event.groupId,
        event.mqType,
        JSON.stringify(event.data),
        event.latencyMs,
    ]);
}

async function runWorker() {
    // 1. DB 연결
    const dbClient = new Client({ connectionString: DB_URL });
    await dbClient.connect();
    console.log(`[BullMQ] Connected to DB for ${GROUP_ID}`);

    // 2. Bull Board UI 서버 시작
    const ordersQueue = new Queue(QUEUE_NAME, { connection: redisConnection });
    const serverAdapter = new ExpressAdapter();
    serverAdapter.setBasePath('/ui');

    createBullBoard({
        queues: [new BullMQAdapter(ordersQueue)],
        serverAdapter,
    });

    const app = express();
    app.use('/ui', serverAdapter.getRouter());
    app.listen(BULL_BOARD_PORT, () => {
        console.log(`[BullMQ] Bull Board UI running at http://localhost:${BULL_BOARD_PORT}/ui`);
    });

    // 3. BullMQ Worker 시작
    const worker = new Worker(
        QUEUE_NAME,
        async (job) => {
            const eventData = job.data as Record<string, unknown>;
            const orderId = (eventData['order_id'] ?? eventData['orderId']) as string;
            const publishedAtStr = (eventData['published_at'] ?? eventData['publishedAt']) as string;

            const consumedAt = new Date();
            const publishedAt = new Date(publishedAtStr);
            const latencyMs = consumedAt.getTime() - publishedAt.getTime();

            console.log(` [x] ${GROUP_ID} Received job #${job.id}: ${orderId} (attempt ${job.attemptsMade + 1})`);

            // SIMULATE_FAILURE: 50% 확률로 강제 실패 → BullMQ가 자동 retry
            if (SIMULATE_FAILURE && Math.random() < 0.5) {
                throw new Error(`[SIMULATE_FAILURE] Intentional failure for job #${job.id}`);
            }

            await insertProcessedEvent(dbClient, {
                eventId: orderId,
                groupId: GROUP_ID,
                mqType: 'bullmq',
                data: eventData,
                latencyMs,
            });

            console.log(` [✓] ${GROUP_ID} Completed job #${job.id}: ${orderId}`);
        },
        {
            connection: redisConnection,
            concurrency: 1,
            // 재시도 횟수·backoff 전략은 Job enqueue 시 opts에 포함됨
            // (Producer의 BullMQProducer.publish에서 설정)
        }
    );

    worker.on('failed', (job, err) => {
        if (job) {
            console.error(` [!] Job #${job.id} failed (attempt ${job.attemptsMade}/${job.opts.attempts ?? 3}): ${err.message}`);
        }
    });

    worker.on('error', (err) => {
        console.error('[BullMQ] Worker error:', err);
    });

    console.log(`[BullMQ] Worker started. Listening on queue "${QUEUE_NAME}" (SIMULATE_FAILURE=${SIMULATE_FAILURE})`);
}

runWorker().catch((err) => {
    console.error('Error running BullMQ Worker:', err);
    process.exit(1);
});

File: workers/python/bullmq_worker.py
"""Python RQ Worker for BullMQ spec.

RQ(Redis Queue)를 사용하여 'orders-python' 큐에서 Job을 가져와 처리합니다.
API Server가 enqueue한 Job을 소비하고 processed_events 테이블에 저장합니다.

Job 처리 함수(process_order)는 module 경로로 참조되므로,
이 파일이 Python path에 있어야 합니다 (workers/python/ 에서 실행).

macOS 실행 시 주의:
    RQ는 내부적으로 fork()를 사용하며, macOS에서 Objective-C 런타임과
    충돌할 수 있습니다. 아래 환경변수를 설정해야 합니다:
    OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES venv/bin/python bullmq_worker.py
"""

import json
import os
import random
import sys
from datetime import datetime, timezone

import redis
from rq import Queue, Worker
from sqlmodel import Session, create_engine

sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'packages'))
from shared_python import ProcessedEvent

DB_URL = os.getenv("DATABASE_URL", "postgresql://postgres:password@localhost:5432/mq_db")
REDIS_URL = os.getenv("REDIS_URL", "redis://localhost:6379")
QUEUE_NAME = "orders-python"
GROUP_ID = "payment-group"
SIMULATE_FAILURE = os.getenv("SIMULATE_FAILURE", "false") == "true"


def _get_engine():
    """RQ 포크 서브프로세스에서 SQLAlchemy 엔진 충돌을 방지하기 위해 lazy 생성."""
    return create_engine(os.getenv("DATABASE_URL", "postgresql://postgres:password@localhost:5432/mq_db"))


def process_order(event_dict: dict) -> dict:
    """RQ Worker가 실행하는 Job 처리 함수.

    API Server가 'bullmq_worker.process_order' 경로로 enqueue합니다.
    RQ Worker는 이 함수를 import하여 실행합니다.
    """
    order_id = event_dict.get("order_id")
    published_at_str = event_dict.get("published_at")

    consumed_at = datetime.now(timezone.utc)
    published_at = datetime.fromisoformat(str(published_at_str).replace("Z", "+00:00"))
    latency_ms = int((consumed_at - published_at).total_seconds() * 1000)

    # SIMULATE_FAILURE: 50% 확률로 강제 실패 → RQ가 자동 retry
    if SIMULATE_FAILURE and random.random() < 0.5:
        raise Exception(f"[SIMULATE_FAILURE] Intentional failure for order {order_id}")

    print(f" [x] {GROUP_ID} Received: {order_id} (Latency: {latency_ms}ms)")

    with Session(_get_engine()) as session:
        event_record = ProcessedEvent(
            event_id=order_id,
            group_id=GROUP_ID,
            mq_type="bullmq",
            data=json.dumps(event_dict),
            latency_ms=latency_ms,
        )
        session.add(event_record)
        session.commit()

    print(f" [✓] {GROUP_ID} Completed: {order_id}")
    return {"status": "ok", "order_id": order_id}


if __name__ == "__main__":
    conn = redis.from_url(REDIS_URL)
    queue = Queue(QUEUE_NAME, connection=conn)
    worker = Worker([queue], connection=conn)

    print(f"[*] Python RQ Worker ({GROUP_ID}) started. Listening on queue '{QUEUE_NAME}' (SIMULATE_FAILURE={SIMULATE_FAILURE})")
    worker.work(with_scheduler=True)
